src/source/query_features.js

```diff
@@ -49,21 +51,38 @@ function sortTilesIn(a, b) {
 }
 
 function mergeRenderedFeatureLayers(tiles) {
-    const result = tiles[0] || {};
-    for (let i = 1; i < tiles.length; i++) {
-        const tile = tiles[i];
-        for (const layerID in tile) {
-            const tileFeatures = tile[layerID];
+    // Avoid merge work for common cases
+    if (!tiles.length) {
+        return {};
+    } else if (tiles.length === 1) {
+        return tiles[0].queryResults;
+    }
+
+    // Merge results from all tiles, but if two tiles share the same
+    // wrapped ID, don't duplicate features between the two tiles
+    const result = {};
+    const wrappedIDFeatureMap = {};
+    for (const tile of tiles) {
+        const queryResults = tile.queryResults;
+        const wrappedID = tile.wrappedTileID;
+        const wrappedIDFeatures = wrappedIDFeatureMap[wrappedID] = wrappedIDFeatureMap[wrappedID] || {};
+        for (const layerID in queryResults) {
+            const tileFeatures = queryResults[layerID];
             let resultFeatures = result[layerID];
             if (resultFeatures === undefined) {
                 resultFeatures = result[layerID] = tileFeatures;
+                tileFeatures.forEach((tileFeature) => {
+                    wrappedIDFeatures[tileFeature.id] = true;
+                });
             } else {
-                for (let f = 0; f < tileFeatures.length; f++) {
-                    resultFeatures.push(tileFeatures[f]);
+                for (const tileFeature of tileFeatures) {
+                    if (!tileFeature.id || !wrappedIDFeatures[tileFeature.id]) {
+                        wrappedIDFeatures[tileFeature.id] = true;
```

Comment:
If the features don't have IDs, they'll get duplicated (because of the `!tileFeature.id` check). If the feature has an ID, it should get included once. Say the new feature has ID "1" and `wrappedIDFeatures` looks like `{ 'undefined': true }`. `wrappedIDFeatures[1]` will still be `undefined`. Am I missing something?

Do you have any opinion on whether it might make sense to just change the behavior here to "duplicate on wrapped tiles"? The current behavior is publicly documented (https://www.mapbox.com/mapbox-gl-js/api/#map#queryrenderedfeatures) so I'm thinking the answer is "no", but on the other hand users of the API can already see the same feature returned multiple times across tile boundaries.